RecoTracker/TkDetLayers/src/GeometricSearchTracker.cc

```diff
@@ -78,46 +78,100 @@ GeometricSearchTracker::~GeometricSearchTracker() {
 
 const DetLayer* GeometricSearchTracker::idToLayer(const DetId& id) const {
-  switch (id.subdetId()) {
-    case StripSubdetector::TIB:
-      return theTibLayers[theTrkTopo->tibLayer(id) - 1];
-      break;
-
-    case StripSubdetector::TOB:
-      return theTobLayers[theTrkTopo->tobLayer(id) - 1];
-      break;
-
-    case StripSubdetector::TID:
-      if (theTrkTopo->tidSide(id) == 1) {
-        return theNegTidLayers[theTrkTopo->tidWheel(id) - 1];
-      } else if (theTrkTopo->tidSide(id) == 2) {
-        return thePosTidLayers[theTrkTopo->tidWheel(id) - 1];
-      }
-      break;
-
-    case StripSubdetector::TEC:
-      if (theTrkTopo->tecSide(id) == 1) {
-        return theNegTecLayers[theTrkTopo->tecWheel(id) - 1];
-      } else if (theTrkTopo->tecSide(id) == 2) {
-        return thePosTecLayers[theTrkTopo->tecWheel(id) - 1];
-      }
-      break;
-
-    case PixelSubdetector::PixelBarrel:
-      return thePixelBarrelLayers[theTrkTopo->pxbLayer(id) - 1];
-      break;
-
-    case PixelSubdetector::PixelEndcap:
-      if (theTrkTopo->pxfSide(id) == 1) {
-        return theNegPixelForwardLayers[theTrkTopo->pxfDisk(id) - 1];
-      } else if (theTrkTopo->pxfSide(id) == 2) {
-        return thePosPixelForwardLayers[theTrkTopo->pxfDisk(id) - 1];
-      }
-      break;
-
-    default:
-      edm::LogError("TkDetLayers") << "ERROR:layer not found!";
-      // throw(something);
+  //If it's a tracker Det
+  if (id.det() == 1) {
+    switch (id.subdetId()) {
+      case StripSubdetector::TIB:
+        return theTibLayers[theTrkTopo->tibLayer(id) - 1];
+        break;
+
+      case StripSubdetector::TOB:
+        return theTobLayers[theTrkTopo->tobLayer(id) - 1];
+        break;
+
+      case StripSubdetector::TID:
+        if (theTrkTopo->tidSide(id) == 1) {
+          return theNegTidLayers[theTrkTopo->tidWheel(id) - 1];
+        } else if (theTrkTopo->tidSide(id) == 2) {
+          return thePosTidLayers[theTrkTopo->tidWheel(id) - 1];
+        }
+        break;
+
+      case StripSubdetector::TEC:
+        if (theTrkTopo->tecSide(id) == 1) {
+          return theNegTecLayers[theTrkTopo->tecWheel(id) - 1];
+        } else if (theTrkTopo->tecSide(id) == 2) {
+          return thePosTecLayers[theTrkTopo->tecWheel(id) - 1];
+        }
+        break;
+
+      case PixelSubdetector::PixelBarrel:
+        return thePixelBarrelLayers[theTrkTopo->pxbLayer(id) - 1];
+        break;
+
+      case PixelSubdetector::PixelEndcap:
+        if (theTrkTopo->pxfSide(id) == 1) {
+          return theNegPixelForwardLayers[theTrkTopo->pxfDisk(id) - 1];
+        } else if (theTrkTopo->pxfSide(id) == 2) {
+          return thePosPixelForwardLayers[theTrkTopo->pxfDisk(id) - 1];
+        }
+        break;
+
+      default:
+        edm::LogError("TkDetLayers") << "ERROR:layer not found!";
+        // throw(something);
+    }
+    return nullptr;  //just to avoid compile warnings
+  } else {
+    //If it's MTD
```

Comment:
Is it safe to assume that "whatever is not Tracker is MTD"? Shouldn't you check for the MTD detId, instead?
(By the way: which is the MTD detID? Is it `DetId::Detector::Forward` also for the BTL?)

Reply:
Hi Andrea,

I have modified this to make sure that it is an MTD detector. (I think it should never happen otherwise, but still is good to check, thanks for the comment).